src/ConnectAuth.ts

```diff
@@ -0,0 +1,192 @@
+  clientKey: string;
+  connectJwt: ConnectJwt;
+  storedEntity: E;
+}
+
+export type VerifyInstallResponse<E> = VerifyInstallNewResponse | VerifyInstallUpdateResponse<E>;
+
+export interface VerifyRequestArgs<E> extends CommonVerifyArgs<E> {
+  useContextJwt?: boolean;
+}
+
+export interface VerifyRequestResponse<E> {
+  connectJwt: ConnectJwt;
+  storedEntity: E;
+}
+
+export interface VerifyQueryStringHashArgs {
+  baseUrl: string;
+  connectJwt: ConnectJwt;
+  requestReader: RequestReader;
+  useContextJwt?: boolean;
+  skipQueryStringHashCheck?: boolean; // Bitbucket does not implement Query String Hash
+}
+
+/**
+ * Entry point for Connect request authentication.
+ */
+export class ConnectAuth {
+  /**
+   * Verifies a Connect request installation.
+   * Use this method to make sure the request is valid before persisting any data.
+   * This method handles both new installations and re-installations or installation updates.
+   */
+  static async verifyInstall<E>({
+    baseUrl,
+    requestReader,
+    loadCredentials,
+    skipQueryStringHashCheck,
+  }: VerifyInstallArgs<E>): Promise<VerifyInstallResponse<E>> {
+    const clientKey = requestReader.extractClientKey();
+
+    // Check issuer
+    const rawConnectJwt = requestReader.extractConnectJwt();
+    let unverifiedConnectJwt;
+    if (rawConnectJwt) {
+      unverifiedConnectJwt = decodeUnverifiedConnectJwt(rawConnectJwt);
+      if (unverifiedConnectJwt.iss !== clientKey) {
+        throw new AuthError('Wrong issuer', {
+          code: AuthErrorCode.WRONG_ISSUER,
+          unverifiedConnectJwt,
+        });
+      }
+    }
+
+    // Check new installation
+    const credentials = await loadCredentials(clientKey);
+    if (!credentials) {
+      return {
+        type: InstallType.newInstall,
+        clientKey,
+      };
+    }
+
+    // Check installation update
+    if (rawConnectJwt) {
+      const connectJwt = verifyConnectJwt({
+        rawConnectJwt,
+        sharedSecret: credentials.sharedSecret,
+        unverifiedConnectJwt,
+      });
+
+      if (!skipQueryStringHashCheck) {
+        ConnectAuth.verifyQsh({ baseUrl, connectJwt, requestReader });
+      }
+
+      return {
+        type: InstallType.update,
+        clientKey,
+        connectJwt,
+        storedEntity: credentials.storedEntity,
+      };
+    }
+
+    throw new AuthError('Unauthorized update request', {
+      code: AuthErrorCode.UNAUTHORIZED_REQUEST,
+    });
+  }
+
+  /**
+   * Verifies any post-installation incoming connect requests using currently stored Shared Secret.
+   * Use this method to verify the request was actually initiated by Atlassian Connect service and that its content
+   * is not tainted via the Query String Hash algorithm.
+   * This method handles both API requests and frame-loading requests, in which case you should pass
+   * `useContextJwt: true`.
+   */
+  static async verifyRequest<E>({
+    baseUrl,
+    requestReader,
+    loadCredentials,
+    useContextJwt,
```

Comment:
I have named this attribute `useContextJwt` instead of `allowContextJwt` because my understanding is the following:

- iframe loading send a JWT whose qsh attribute will be `context-jwt` 
- so the library won't compute a QSH from the request but will *use* a static value `context-qsh` to validate the incoming request.

Reply:
JWT from iframe has a valid QSH. `context-jwt` is only for `window.AP.context.getToken()`.